airflow/executors/dask_executor.py

```diff
@@ -87,19 +90,23 @@ def _process_future(self, future: Future) -> None:
             self.futures.pop(future)
 
     def sync(self) -> None:
-        assert self.futures, NOT_STARTED_MESSAGE
+        if not self.futures:
+            raise AirflowException(NOT_STARTED_MESSAGE)
         # make a copy so futures can be popped during iteration
         for future in self.futures.copy():
             self._process_future(future)
 
     def end(self) -> None:
-        assert self.client, NOT_STARTED_MESSAGE
-        assert self.futures, NOT_STARTED_MESSAGE
+        if not self.client:
```

Comment:
```suggestion
        if not self.client or not self.futures:
```

Reply:
I prefer not to. Those check should be separated out, because if they fail due to programmatic error you should kno which value was null. With or you never know which was null.